File: frontend/src/api/basedata.ts
// 基础数据(教师/科目/教室/场地/班级)API 类型与调用封装。

import { apiGet, apiPost, request } from '@/api/client'

export type RoomType = 'normal' | 'special' | 'workshop' | 'outdoor'
export type ClassTrack = 'elementary' | 'junior_high' | 'senior_high' | 'comprehensive' | 'vocational'
export type TeacherRuleType = 'unavailable' | 'avoid' | 'prefer'

export interface SubjectBrief {
  id: number
  name: string
}
export interface Subject {
  id: number
  semester_id: number
  name: string
  domain: string | null
  required_room_type: RoomType | null
  default_block_size: number
  /** 主科(国英数等):自动排课会尽量把主科排在上午 */
  is_major: boolean
}
export interface Teacher {
  id: number
  semester_id: number
  name: string
  base_periods: number
  admin_title: string | null
  admin_reduction: number
  is_external: boolean
  is_active: boolean
  subjects: SubjectBrief[]
  email: string | null
  phone: string | null
  line_id: string | null
  user_id: number | null
}
export interface BindableAccount {
  id: number
  username: string
  display_name: string
}
export interface TeacherTimeRule {
  id?: number
  weekday: number
  period_no: number
  rule_type: TeacherRuleType
}
export interface Room {
  id: number
  semester_id: number
  name: string
  room_type: RoomType
  capacity: number | null
  subjects: SubjectBrief[]
}
export interface ClassUnit {
  id: number
  semester_id: number
  grade: number
  name: string
  track: ClassTrack
  department: string | null
  student_count: number | null
  homeroom_teacher_id: number | null
  homeroom_teacher: SubjectBrief | null
  period_table_id: number | null
}

export const ROOM_TYPE_LABELS: Record<RoomType, string> = {
  normal: '普通教室',
  special: '专用教室',
  workshop: '实训场地',
  outdoor: '户外',
}
export const TRACK_LABELS: Record<ClassTrack, string> = {
  elementary: '小学',
  junior_high: '初中',
  senior_high: '普通高中',
  comprehensive: '综合高中',
  vocational: '中职',
}
export const RULE_TYPE_LABELS: Record<TeacherRuleType, string> = {
  unavailable: '不可排',
  avoid: '尽量避开',
  prefer: '偏好',
}

// ── 科目 ──
export const listSubjects = (semesterId: number, q?: string) =>
  apiGet<Subject[]>(`/subjects?semester_id=${semesterId}${q ? `&q=${encodeURIComponent(q)}` : ''}`)
export const createSubject = (semesterId: number, body: Partial<Subject>) =>
  apiPost<Subject>(`/subjects?semester_id=${semesterId}`, body)
export const updateSubject = (id: number, body: Partial<Subject>) =>
  request<Subject>('PATCH', `/subjects/${id}`, body)
export const deleteSubject = (id: number) => request<void>('DELETE', `/subjects/${id}`)

// ── 教师 ──
export const listTeachers = (semesterId: number, q?: string) =>
  apiGet<Teacher[]>(`/teachers?semester_id=${semesterId}${q ? `&q=${encodeURIComponent(q)}` : ''}`)
export const createTeacher = (semesterId: number, body: Record<string, unknown>) =>
  apiPost<Teacher>(`/teachers?semester_id=${semesterId}`, body)
export const updateTeacher = (id: number, body: Record<string, unknown>) =>
  request<Teacher>('PATCH', `/teachers/${id}`, body)
export const deleteTeacher = (id: number) => request<void>('DELETE', `/teachers/${id}`)
export const listBindableAccounts = (semesterId: number, currentTeacherId?: number) =>
  apiGet<BindableAccount[]>(
    `/teachers/bindable-accounts?semester_id=${semesterId}` +
      (currentTeacherId ? `&current_teacher_id=${currentTeacherId}` : ''),
  )
export const getTimeRules = (id: number) => apiGet<TeacherTimeRule[]>(`/teachers/${id}/time-rules`)
export const replaceTimeRules = (id: number, rules: TeacherTimeRule[]) =>
  request<TeacherTimeRule[]>('PUT', `/teachers/${id}/time-rules`, rules)

// ── 教室/场地 ──
export const listRooms = (semesterId: number, q?: string) =>
  apiGet<Room[]>(`/rooms?semester_id=${semesterId}${q ? `&q=${encodeURIComponent(q)}` : ''}`)
export const createRoom = (semesterId: number, body: Record<string, unknown>) =>
  apiPost<Room>(`/rooms?semester_id=${semesterId}`, body)
export const updateRoom = (id: number, body: Record<string, unknown>) =>
  request<Room>('PATCH', `/rooms/${id}`, body)
export const deleteRoom = (id: number) => request<void>('DELETE', `/rooms/${id}`)

// ── 班级 ──
export const listClassUnits = (semesterId: number, q?: string) =>
  apiGet<ClassUnit[]>(
    `/class-units?semester_id=${semesterId}${q ? `&q=${encodeURIComponent(q)}` : ''}`,
  )
export const createClassUnit = (semesterId: number, body: Record<string, unknown>) =>
  apiPost<ClassUnit>(`/class-units?semester_id=${semesterId}`, body)
export const updateClassUnit = (id: number, body: Record<string, unknown>) =>
  request<ClassUnit>('PATCH', `/class-units/${id}`, body)
export const deleteClassUnit = (id: number) => request<void>('DELETE', `/class-units/${id}`)

File: backend/app/models/basedata.py
"""基础数据 model:教师、科目、教室/场地、班级、教师时段规则。

均隶属于某学期(semester_id 范围,见 D3 学期快照)。
"""

import enum
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    Column,
    ForeignKey,
    Integer,
    String,
    Table,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.db import Base

if TYPE_CHECKING:
    from app.models.user import User


class RoomType(enum.StrEnum):
    normal = "normal"      # 普通教室
    special = "special"    # 专用教室
    workshop = "workshop"  # 实训场地
    outdoor = "outdoor"    # 户外


class ClassTrack(enum.StrEnum):
    elementary = "elementary"      # 小学
    junior_high = "junior_high"    # 初中
    senior_high = "senior_high"    # 普通高中
    comprehensive = "comprehensive"  # 综合高中
    vocational = "vocational"      # 中职


class TeacherRuleType(enum.StrEnum):
    unavailable = "unavailable"  # 不可排(硬约束)
    avoid = "avoid"              # 尽量避开(软约束)
    prefer = "prefer"            # 偏好(软约束)


# ── 多对多关联表 ──────────────────────
teacher_subjects = Table(
    "teacher_subjects",
    Base.metadata,
    Column("teacher_id", ForeignKey("teachers.id", ondelete="CASCADE"), primary_key=True),
    Column("subject_id", ForeignKey("subjects.id", ondelete="CASCADE"), primary_key=True),
)

room_subjects = Table(
    "room_subjects",
    Base.metadata,
    Column("room_id", ForeignKey("rooms.id", ondelete="CASCADE"), primary_key=True),
    Column("subject_id", ForeignKey("subjects.id", ondelete="CASCADE"), primary_key=True),
)


class Subject(Base):
    __tablename__ = "subjects"

    id: Mapped[int] = mapped_column(primary_key=True)
    semester_id: Mapped[int] = mapped_column(
        ForeignKey("semesters.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(64))
    domain: Mapped[str | None] = mapped_column(String(64), nullable=True)  # 领域/群别
    required_room_type: Mapped[str | None] = mapped_column(String(20), nullable=True)
    default_block_size: Mapped[int] = mapped_column(Integer, default=1)  # 默认连堂长度(1=不连堂)
    # 主科(国英数等):排课引擎的软约束 S5 会尽量把主科排在上午
    is_major: Mapped[bool] = mapped_column(Boolean, default=False)


class Teacher(Base):
    __tablename__ = "teachers"
    # 一个账号在同一学期至多绑定一位教师(user_id 为空时不受限,见 M2-0)
    __table_args__ = (
        UniqueConstraint("semester_id", "user_id", name="uq_teachers_semester_user"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    semester_id: Mapped[int] = mapped_column(
        ForeignKey("semesters.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(32))
    id_last4: Mapped[str | None] = mapped_column(String(4), nullable=True)  # 身份证后四位(辅助识别)
    base_periods: Mapped[int] = mapped_column(Integer, default=0)  # 基本课时
    admin_title: Mapped[str | None] = mapped_column(String(32), nullable=True)  # 行政职称
    admin_reduction: Mapped[int] = mapped_column(Integer, default=0)  # 行政减课节数
    is_external: Mapped[bool] = mapped_column(Boolean, default=False)  # 外聘/企业兼职教师
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)  # 在职
    # 联系信息(均选填,供调课与代课通知与人工联系;挂教师因外聘教师可能无系统账号)
    email: Mapped[str | None] = mapped_column(String(128), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(32), nullable=True)
    line_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    # 绑定的登录账号(空=无账号,如外聘教师);账号删除时设为 NULL 保留教师基础信息
    user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    subjects: Mapped[list[Subject]] = relationship(secondary=teacher_subjects, lazy="selectin")
    time_rules: Mapped[list["TeacherTimeRule"]] = relationship(
        back_populates="teacher", cascade="all, delete-orphan", lazy="selectin"
    )
    user: Mapped["User | None"] = relationship(lazy="selectin")


class TeacherTimeRule(Base):
    __tablename__ = "teacher_time_rules"
    __table_args__ = (
        UniqueConstraint("teacher_id", "weekday", "period_no", name="uq_teacher_time_rules_cell"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    teacher_id: Mapped[int] = mapped_column(
        ForeignKey("teachers.id", ondelete="CASCADE"), index=True
    )
    weekday: Mapped[int] = mapped_column(Integer)
    period_no: Mapped[int] = mapped_column(Integer)
    rule_type: Mapped[str] = mapped_column(String(20))

    teacher: Mapped[Teacher] = relationship(back_populates="time_rules")


class Room(Base):
    __tablename__ = "rooms"

    id: Mapped[int] = mapped_column(primary_key=True)
    semester_id: Mapped[int] = mapped_column(
        ForeignKey("semesters.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(64))
    room_type: Mapped[str] = mapped_column(String(20), default=RoomType.normal.value)
    capacity: Mapped[int | None] = mapped_column(Integer, nullable=True)

    subjects: Mapped[list[Subject]] = relationship(secondary=room_subjects, lazy="selectin")


class ClassUnit(Base):
    __tablename__ = "class_units"
    # 同学期班名唯一(M6-5):冲突信息、课表、导出全都以班名指称班级,
    # 同学期两个「301」会让排课管理员在页面上分不出是哪一班。
    __table_args__ = (
        UniqueConstraint("semester_id", "name", name="uq_class_units_semester_name"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    semester_id: Mapped[int] = mapped_column(
        ForeignKey("semesters.id", ondelete="CASCADE"), index=True
    )
    grade: Mapped[int] = mapped_column(Integer)       # 年级
    name: Mapped[str] = mapped_column(String(32))     # 班名
    track: Mapped[str] = mapped_column(String(20))    # 学制标签
    department: Mapped[str | None] = mapped_column(String(32), nullable=True)  # 专业类别(中职)
    student_count: Mapped[int | None] = mapped_column(Integer, nullable=True)  # 人数
    homeroom_teacher_id: Mapped[int | None] = mapped_column(
        ForeignKey("teachers.id", ondelete="SET NULL"), nullable=True, index=True
    )
    # 所属作息时间表(空=用学期默认表);支持完全中学/附设小学部等多套作息时间表场景
    period_table_id: Mapped[int | None] = mapped_column(
        ForeignKey("period_tables.id", ondelete="SET NULL"), nullable=True, index=True
    )

    homeroom_teacher: Mapped[Teacher | None] = relationship(lazy="selectin")
